File: x10.lib/src/sched/Pool.cc
/*
============================================================================
 Name        : Pool.cpp
 Author      : Rajkishore Barik
 Version     :
 Copyright   : IBM Corporation 2007
 Description : Exe source file
============================================================================
*/
using namespace x10lib_cws;
using namespace std;
#include <time.h>

Pool::~Pool() {
	pthread_cond_destroy(&work);
	pthread_cond_destroy(&termination);
	delete lock;
	delete barrier;
	free id;
	for(i=0;i<workers.size();i++)
					delete workers[i];
	workers.clear();
}


void Pool::each_thread(int d)
{
	 Worker *ws = new Worker(d, this);
	 workers[d]=ws;
     //barrier->barrier();
     /*if (id == 0)
	  ws->run(invoke_main);
     else
      ws->run((Closure *) NULL);*/
     ws->run();
}

void Pool::barrierAction(Closure *c)
{
	if (c != NULL && c->requiresGlobalQuiescence()) 
			c->completed();
	c = NULL;
	MEM_BARRIER(); // let everyone see it -- TODO RAJ
}

Pool::Pool(int numThreads) {
	
	long i;
	int res;
	pthread_attr_t attr;
	
	if (numThreads <=0) { cout << "Illegal argument"; abort(); }
	
	lock = new PosixLock();

	pthread_cond_init (&work, NULL);
	pthread_cond_init (&termination, NULL);

	barrier = new ActiveWorkerCount(&barrierAction, currentJob);

	num_workers = numThreads;
	runningWorkers = 0;
	activeJobs = 0;
	activeOnJobAtomic = 0;
	jobs =  new JobQueue();
	joinCount = 0;
	
	workers.resize(INITIAL_WORKER_CAPACITY);
	
	  
	id = malloc(num_workers * sizeof(pthread_t));
	

	pthread_attr_init(&attr); 
	pthread_attr_setscope(&attr, PTHREAD_SCOPE_SYSTEM); 
	 
	lock -> lock_wait_posix();
	  
	for (i = 0; i < num_workers; i++)
	{
		  res = pthread_create(id + i, 
				       &attr,
				       (void * (*) (void *)) each_thread, 
				       (void *) i);
		  if (res) {cout << "could not create"; abort(); }
		  runningWorkers++;
		       
	}
	lock -> lock_signal_posix();
}

/*
Pool::shutDown() {

	for (i = 0; i < num_workers; i++)
	{
		  res = pthread_join(id[i], NULL);
		  if (res) cout << "could not join";
	}
}*/


/**
     * Sets the handler for internal worker threads that terminate due
     * to uncaught Errors or other unchecked Throwables encountered
     * while executing tasks. Since such errors are not in generable
     * recoverable, they are not managed by ForkJoinTasks themselves,
     * but instead cause threads to die, invoking the handler, and
     * then being replaced.  In many usage contexts, it is probably a
     * better ideas to install a handler to either shutdown the pool
     * or exit the program when such errors occur.  Unless set, the
     * current default or ThreadGroup handler is used as handler.
     * @param h the new handler
     * @return the old handler, or null if none
     */
/*
UncaughtExceptionHandler *Pool::setUncaughtExceptionHandler(UncaughtExceptionHandler *h) {
        UncaughtExceptionHandler *old = NULL;
				abort(); // TODO RAJ
        const PosixLock *l = this->lock;
        l->lock_wait_posix();
        old = ueh;
        ueh = h;
        for (int i = 0; i < workers.size(); ++i) {
        	Worker *w = workers[i];
            if (w != NULL)
                    w.setUncaughtExceptionHandler(h);
            
        }
        l->lock_signal_posix();
        return old;
}
*/

int Pool::getPoolSize() const {
        return workers.size();
}
int Pool::activeOnJob() const {
    	return activeOnJobAtomic;
}

bool Pool::isShutdown() const {
        return runState >= SHUTDOWN;
}

bool Pool::isTerminated() const {
        return runState == TERMINATED;
}

void Pool::initFrameGenerator(FrameGenerator *fg) {
        for (int i = 0; i < workers.size(); ++i) {
            Worker *t = workers[i];
            t->setFrameGenerator(fg);
        }
}

   /**
     * Initiates an orderly shutdown in which previously submitted
     * tasks are executed, but no new tasks will be accepted.
     * Invocation has no additional effect if already shut down.
     */
void Pool::shutdown() {
        // todo security checks??
        PosixLock *l = this->lock;
        l->lock_wait_posix();
        
        if (runState < SHUTDOWN) {
                   runState = SHUTDOWN;
									 MEM_BARRIER(); // for volatile update -- TODO RAJ
        }
        l->lock_signal_posix();
}

    /**
     * Attempts to stop all actively executing tasks, and halts the
     * processing of waiting tasks.
     */
void Pool::shutdownNow() {
        PosixLock *l = this->lock;
        l->lock_wait_posix();
        //tryTerminate 
        l->lock_signal_posix();
}

// RAJ:: timeout is used as it is without conversion
bool Pool::awaitTermination(const struct timespec *timeout) {
        
        PosixLock *l = this->lock;
        l->lock_wait_posix();
        
        for (;;) {
                if (runState == TERMINATED)
                    return true;
                if (timeout->tv_nsec <= 0)
                    return false;
                // nanos = termination.awaitNanos(timeout); // TODO wait for timeout nanos RAJ
								pthread_cond_timedwait(&termination,this->lock->posix_lock_var,timeout);

            }
        l->lock_signal_posix();
}
// Internal methods that may be invoked by workers

bool Pool::isStopped() const {
     return runState >= STOP;
}

 /**
  * Return the workers array; needed for random-steal by Workers.
  */
vector<Worker *> *Pool::getWorkers() const {
     return &workers;
}
 
void Pool::submit(Job *job) {
    	//startTime = System.nanoTime();
        addJob(job);
}
void Pool::invoke(Job *job) {
    	submit(job);
    	//try {
    	job->waitForCompletion();
		//} catch (InterruptedException z) {} // TODO
}

/**
     * Enqueue an externally submitted task
     */
void Pool::addJob(Job *job) {
        PosixLock *l = this->lock;
        bool ok;
        l->lock_wait_posix();
        
        if (ok = (runState == RUNNING)) {
                jobs->add(job);
                //work.signalAll();
                pthread_cond_broadcast(&work);
        }
        
        l->lock_signal_posix();
        
        if (!ok) { abort(); }
            //throw RejectedExecutionException();
}

/**
     * Returns the total number of tasks stolen from one thread's work
     * queue by another. This value is only an approximation,
     * obtained by iterating across all threads in the pool, but may
     * be useful for monitoring and tuning fork/join programs.
     * @return the number of steals.
     */
long Pool::getStealCount() {
        long sum = 0;
        for (int i = 0; i < workers.size(); ++i) {
            Worker *t = workers[i];
            if (t != NULL) 
                sum += t->stealCount;
        }
        return sum;
}
    /**
     * Returns the total number of steal attempts made by all workers.
     * This value is only an approximation,
     * obtained by iterating across all threads in the pool, but may
     * be useful for monitoring and tuning fork/join programs.
     * @return the number of steal attempts.
     */
long Pool::getStealAttempts() {
    	long sum = 0;
      for (int i = 0; i < workers.size(); ++i) {
            Worker *t = workers[i];
            if (t != NULL) 
                sum += t->stealAttempts;
      }
      return sum;
}
    /**
     * Termination callback from dying worker.
     */
void Pool::workerTerminated(Worker *r, int index) {
        PosixLock *l = this->lock;
        l->lock_wait_posix();
        
        if (runState >= STOP) {
        	if (--runningWorkers <= 0) {
        		runState = TERMINATED;
						MEM_BARRIER(); // TODO RAj
                    //termination.signalAll();
        		pthread_cond_broadcast(&termination);
            }
        }
        else if (index >= 0 && index < workers.size() &&
                     workers[index] == r) {
        	abort();
        	/*// TODO Confirm from Vijay
            Worker *replacement = new Worker(this, index);
            if (ueh != NULL)
                    replacement.setUncaughtExceptionHandler(ueh);
                workers[index] = replacement;
                replacement.start();
                */
         }
        
         l->lock_signal_posix();
}

void Pool::jobCompleted() {
        PosixLock *l = this->lock;
        l->lock_wait_posix();
        
        if (--activeJobs <= 0 && runState == SHUTDOWN && jobs->isEmpty())
                tryTerminate();
        
        l->lock_signal_posix();
        
}
void Pool::tryTerminate() const {
    	// do nothing.
}
Closure *Pool::getJob() const {
        
        lock->lock_wait_posix();
        //try {
        	Closure *task = jobs->poll();
        	if (task == NULL && activeJobs == 0) {
        		//work.await(); // TODO RAJ
        		pthread_cond_wait(&work,&(this->lock->posix_lock_var))
        		task=jobs->poll();
        	}
        	if (task != NULL) {
        		++ activeJobs;
        		currentJob = task;
						MEM_BARRIER(); // Let the update be visible to everyone -- TODO RAJ
        		//work.signalAll(); // TODO RAJ
        		pthread_cond_broadcast(&work);
        	}
        //} catch(InterruptedException e) { return NULL; }
        lock->lock_signal_posix();
        return task;
}



JobQueue::JobQueue() {
	elements.resize(INITIAL_JOBQUEUE_CAPACITY);
}
JobQuque::~JobQueue() {
				elements.clear();
}

bool JobQueue::isEmpty() {
	return head == tail;
}

void JobQueue::add(Job *e) {
        elements[tail] = e;
        if ( (tail = (tail + 1) & (elements.size() - 1)) == head)
            doubleCapacity();
}

Job *JobQueue::poll() {
        int h = head;
        Job *result = elements[h]; 
        if (result != NULL) {
            elements[h] = NULL;
            head = (h + 1) & (elements.size() - 1);
        }
        return result;
}
   
void JobQueue::doubleCapacity() {
				int i;
        int p = head;
        int n = elements.size();
        int r = n - p; 
        int newCapacity = n << 1;
        if (newCapacity < 0) { abort();} // TODO throw exception if needed }
        elements.resize(newCapacity); 

				// TODO RAJ: please optimize this code later
				vector<Job *> tmp(INITIAL_JOBQUEUE_CAPACITY);
        for(i=p; i<n; i++) tmp[i-p]=elements[i];
				for(i=0;i<p;i++) tmp[r+i] = elements[i];
				for(i=0;i<n;i++) elements[i] = tmp[i];
				tmp.clear();
        head = 0;
        tail = n;
}
bool JobQueue::isQuiescent() {
        for (int i = 0; i < workers.size(); ++i) {
            Worker *t = workers[i];
            if (t != NULL && t->isActive())
                return false;
        }
        return true;
}

File: tags/pre-june-2009-reorg/x10.runtime.cws/src/x10/runtime/cws/Pool.java

/*
 * A pool of workers for executing X10 programs.
 * 
 * Large portions of code adapted from Doug Lea's jsr166y library, which code
 * carries the header: 
 * Written by Doug Lea with assistance from members of JCP JSR-166
 * Expert Group and released to the public domain, as explained at
 * http://creativecommons.org/licenses/publicdomain
 * 
 * The design of this library is based on the Cilk runtime, developed by the Cilk
 * group at MIT.
 * 
 */
package x10.runtime.cws;

import java.util.concurrent.CyclicBarrier;
import java.util.concurrent.RejectedExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.locks.Condition;
import java.util.concurrent.locks.ReentrantLock;

/**
 * The host and external interface for ForkJoinTasks. A ForkJoinPool
 * manages a group of specialized worker threads that perform
 * ForkJoinTasks. It also provides the entry point for tasks submitted
 * from non-ForkJoinTasks.
 *
 * <p> Class ForkJoinPool does not implement the ExecutorService
 * interface because it only executes ForkJoinTasks, not arbitrary
 * Runnables. However, for the sake of uniformity, it supports all
 * ExecutorService lifecycle control methods (such as shutdown).
 */
public class Pool {
    /*
     * This is an overhauled version of the framework described in "A
     * Java Fork/Join Framework" by Doug Lea, in, Proceedings, ACM
     * JavaGrande Conference, June 2000
     * (http://gee.cs.oswego.edu/dl/papers/fj.pdf). It retains most of
     * the basic structure, but includes a number of algorithmic
     * improvements, along with integration with other
     * java.util.concurrent components.
     */

    /**
     * Main lock protecting access to threads and run state. You might
     * think that having a single lock and condition here wouldn't
     * work so well. But serializing the starting and stopping of
     * worker threads (its main purpose) helps enough in controlling
     * startup effects, contention vs dynamic compilation, etc to be
     * better than alternatives.
     */
    private final ReentrantLock lock = new ReentrantLock();

    /**
     * Condition triggered when new work is available so workers
     * should awaken if blocked. Also used for timed waits when work
     * is known to be available but repeated steal attempts fail
     */
    private final Condition work = lock.newCondition();

    /**
     * Tracks whether pool is running, shutdown, etc. Modified
     * only under lock, but volatile to allow concurrent reads.
     */
    private volatile int runState;

    // Values for runState
    static final int RUNNING    = 0;
    static final int SHUTDOWN   = 1;
    static final int STOP       = 2;
    static final int TERMINATED = 3;
  
    AtomicInteger joinCount = new AtomicInteger();

    /**
     * The pool of threads. Currently, all threads are created
     * upon construction. However, all usages of workers array
     * are prepared to see null entries, allowing alternative
     * schemes in which workers are added more lazily.
     */
     final Worker[] workers;

    
    /**
     * The number of jobs that are currently executing in the pool.
     * This does not include those jobs that are still in job queue
     * waiting to be taken by workers.
     */
    int activeJobs = 0;

    /** The number of workers that have not yet terminated */
    private int runningWorkers = 0;

    /**
     * Condition for awaitTermination. Triggered when
     * activeWorkers reaches zero.
     */
    private final Condition termination = lock.newCondition();

    /**
     * The uncaught exception handler used when any worker
     * abrupty terminates
     */
    private volatile Thread.UncaughtExceptionHandler ueh;
    
    /**
     * The current job being executed by workers in the pool.
     * Set by the worker that retrieves the job from the pool.
     * Unset by the worker that detects quiescence.
     */
    volatile Job currentJob;

    ActiveWorkerCount barrier;
    
    /**
     * Creates a ForkJoinPool with a pool size equal to the number of
     * processors available on the system.
     */
    public Pool() {
        this(Runtime.getRuntime().availableProcessors());
        
    }

    long time() {
    	if (currentJob==null) return 0L;
    	return ((System.nanoTime() - currentJob.startTime)/1000000);
    }
    
    /**
     * Creates a ForkJoinPool with the indicated number
     * of Worker threads.
     */
    public interface WorkerMaker {
    	Worker makeWorker(Pool p, int index);
    }
    public static final WorkerMaker defaultMaker = new WorkerMaker() { 
    	public Worker makeWorker(Pool p, int index) {
    		return new Worker(p,index);
    	}
    };
    public Pool(int poolSize) {
    	this(poolSize, defaultMaker);
    }
    public Pool(final int poolSize, final WorkerMaker m) {
        if (poolSize <= 0) throw new IllegalArgumentException();
        Worker.workers = workers = new Worker[poolSize];
        barrier = new ActiveWorkerCount(new Runnable() { 
        	public void run() {
        		
        		Job job = currentJob;
        		//System.out.println();
        		//System.out.println(Thread.currentThread() + " in barrier w/ " + job);
        		currentJob=null;
        		//printStats();
        		if (job != null && job.requiresGlobalQuiescence()) {
        			job.completed();
        		}
        		
        		
        	}
        }, poolSize);
        lock.lock();
        try {
            for (int i = 0; i < poolSize; ++i) {
                Worker r = m.makeWorker(this, i);
                workers[i] = r;
            }
            for (int i = 0; i < poolSize; ++i) {
            	workers[i].start();
            	++runningWorkers;
            }
            
        } finally {
            lock.unlock();
        }
    }
  
    /**
     * Sets the handler for internal worker threads that terminate due
     * to uncaught Errors or other unchecked Throwables encountered
     * while executing tasks. Since such errors are not in generable
     * recoverable, they are not managed by ForkJoinTasks themselves,
     * but instead cause threads to die, invoking the handler, and
     * then being replaced.  In many usage contexts, it is probably a
     * better ideas to install a handler to either shutdown the pool
     * or exit the program when such errors occur.  Unless set, the
     * current default or ThreadGroup handler is used as handler.
     * @param h the new handler
     * @return the old handler, or null if none
     */
    public Thread.UncaughtExceptionHandler setUncaughtExceptionHandler(Thread.UncaughtExceptionHandler h) {
        final ReentrantLock lock = this.lock;
        Thread.UncaughtExceptionHandler old = null;
        lock.lock();
        try {
            old = ueh;
            ueh = h;
            for (int i = 0; i < workers.length; ++i) {
                Worker w = workers[i];
                if (w != null)
                    w.setUncaughtExceptionHandler(h);
            }
        } finally {
            lock.unlock();
        }
        return old;
    }

    /**
     * Returns the number of worker threads in this pool.
     *
     * @return the number of worker threads in this pool
     */
    public int getPoolSize() {
        return workers.length;
    }
    AtomicInteger activeOnJob = new AtomicInteger();
    public int activeOnJob() {
    	return activeOnJob.intValue();
    }

    /**
     * Returns <tt>true</tt> if this pool has been shut down.
     *
     * @return <tt>true</tt> if this pool has been shut down
     */
    public boolean isShutdown() {
        return runState >= SHUTDOWN;
    }

    /**
     * Returns <tt>true</tt> if all tasks have completed following shut down.
     * Note that <tt>isTerminated</tt> is never <tt>true</tt> unless
     * either <tt>shutdown</tt> or <tt>shutdownNow</tt> was called first.
     *
     * @return <tt>true</tt> if all tasks have completed following shut down
     */
    public boolean isTerminated() {
        return runState == TERMINATED;
    }

    /**
     * Initiates an orderly shutdown in which previously submitted
     * tasks are executed, but no new tasks will be accepted.
     * Invocation has no additional effect if already shut down.
     */
    public void shutdown() {
        // todo security checks??
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            if (runState < SHUTDOWN) {
                    runState = SHUTDOWN;
            }
        } finally {
            lock.unlock();
        }
    }

    /**
     * Attempts to stop all actively executing tasks, and halts the
     * processing of waiting tasks.
     */
    public void shutdownNow() {
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
           // tryTerminate();
        } finally {
            lock.unlock();
        }
    }

    /**
     * Blocks until all tasks have completed execution after a shutdown
     * request, or the timeout occurs, or the current thread is
     * interrupted, whichever happens first.
     *
     * @param timeout the maximum time to wait
     * @param unit the time unit of the timeout argument
     * @return <tt>true</tt> if this executor terminated and
     *         <tt>false</tt> if the timeout elapsed before termination
     * @throws InterruptedException if interrupted while waiting
     */
    public boolean awaitTermination(long timeout, TimeUnit unit)
        throws InterruptedException {
        long nanos = unit.toNanos(timeout);
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            for (;;) {
                if (runState == TERMINATED)
                    return true;
                if (nanos <= 0)
                    return false;
                nanos = termination.awaitNanos(nanos);
            }
        } finally {
            lock.unlock();
        }
    }

    // Internal methods that may be invoked by workers

    final boolean isStopped() {
        return runState >= STOP;
    }

    /**
     * Return the workers array; needed for random-steal by Workers.
     */
    final Worker[] getWorkers() {
        return workers;
    }
    
    public  void submit(Job job) {
    	job.startTime = System.nanoTime();
        addJob(job);
    }
    public void invoke(Job job) {
    	submit(job);
    	try {
			job.waitForCompletion();
			currentJob=null;
		} catch (InterruptedException z) {}
		job.completionTime = System.nanoTime();
    }

    /**
     * Enqueue an externally submitted task
     */
    private void addJob(Job job) {
        final ReentrantLock lock = this.lock;
        boolean ok;
        lock.lock();
        try {
            if (ok = (runState == RUNNING)) {
                jobs.add(job);
                work.signalAll();
            }
        } finally {
            lock.unlock();
        }
        if (!ok)
            throw new RejectedExecutionException();
    }

    /**
     * Returns the total number of tasks stolen from one thread's work
     * queue by another. This value is only an approximation,
     * obtained by iterating across all threads in the pool, but may
     * be useful for monitoring and tuning fork/join programs.
     * @return the number of steals.
     */
    public long getStealCount() {
        long sum = 0;
        for (int i = 0; i < workers.length; ++i) {
            Worker t = workers[i];
            if (t != null) 
                sum += t.stealCount;
        }
        return sum;
    }
    /**
     * Returns the total number of steal attempts made by all workers.
     * This value is only an approximation,
     * obtained by iterating across all threads in the pool, but may
     * be useful for monitoring and tuning fork/join programs.
     * @return the number of steal attempts.
     */
    public long getStealAttempts() {
    	long sum = 0;
        for (int i = 0; i < workers.length; ++i) {
            Worker t = workers[i];
            if (t != null) 
                sum += t.stealAttempts;
        }
        return sum;
    }
    public void printStats() {
    	long stealCount = getStealCount();
    	
    	System.out.print("stealCount=" + stealCount + " ");
    	for (Worker w : workers) {
    		//System.out.println(w + ".stealCount=" + w.stealCount ); 
    		w.stealCount=0;
    	}
    }
    
    public void initFrameGenerator(Worker.FrameGenerator fg) {
        for (int i = 0; i < workers.length; ++i) {
            Worker t = workers[i];
            t.setFrameGenerator(fg);
        }
    }
    /**
     * Termination callback from dying worker.
     */
    final void workerTerminated(Worker r, int index) {
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            if (runState >= STOP) {
                if (--runningWorkers <= 0) {
                    runState = TERMINATED;
                    termination.signalAll();
                }
            }
            else if (index >= 0 && index < workers.length &&
                     workers[index] == r) {
                Worker replacement = new Worker(this, index);
                if (ueh != null)
                    replacement.setUncaughtExceptionHandler(ueh);
                workers[index] = replacement;
                replacement.start();
            }
        } finally {
            lock.unlock();
        }
    }
    
    final void jobCompleted() {
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
            if (--activeJobs <= 0 && runState == SHUTDOWN && jobs.isEmpty())
                tryTerminate();
        } finally {
            lock.unlock();
        }
    }
    public void tryTerminate() {
    	// do nothing.
    }
    
    /**
     * External submission queue.  "Jobs" are tasks submitted to the
     * pool, not internally generated.
     */
    private final JobQueue jobs = new JobQueue();
    /**
     * Returns a job to run, or null if none available.
     */
    final Job getJob() {
        final ReentrantLock lock = this.lock;
        lock.lock();
        try {
        	Job task = jobs.poll();
        	if (task == null && activeJobs ==0) {
        		work.await();
        		task=jobs.poll();
        	}
        	if (task !=null) {
        		++ activeJobs;
        		currentJob = task;
        		work.signalAll();
        	}
        	return task;
        } catch (InterruptedException e) {
        		return null; // ignore interrupt.
        } finally {
        	lock.unlock();
        }
    }
    /**
     * A JobQueue is a simple array-based circular queue.
     * Basically a stripped-down variant of ArrayDeque
     */
    static final class JobQueue {
        static final int INITIAL_JOBQUEUE_CAPACITY = 64;
        Job[] elements = (Job[]) new Job[INITIAL_JOBQUEUE_CAPACITY];
        int head;
        int tail;

        boolean isEmpty() {
            return head == tail;
        }

        void add(Job e) {
            elements[tail] = e;
            if ( (tail = (tail + 1) & (elements.length - 1)) == head)
                doubleCapacity();
        }

        Job poll() {
            int h = head;
            Job result = elements[h]; 
            if (result != null) {
                elements[h] = null;
                head = (h + 1) & (elements.length - 1);
            }
            return result;
        }
        
        void doubleCapacity() {
            int p = head;
            int n = elements.length;
            int r = n - p; 
            int newCapacity = n << 1;
            if (newCapacity < 0)
                throw new IllegalStateException("Job queue capacity exceeded");
            Job[] a = (Job[]) new Job[newCapacity];
            System.arraycopy(elements, p, a, 0, r);
            System.arraycopy(elements, 0, a, r, p);
            elements = a;
            head = 0;
            tail = n;
        }
    }
    /**
     * Returns true if all threads are currently idle.
     * @return true is all threads are currently idle
     */
    public boolean isQuiescent() {
        for (int i = 0; i < workers.length; ++i) {
            Worker t = workers[i];
            if (t != null && t.isActive())
                return false;
        }
        return true;
    }
}
